### PR TITLE
Allow for changing runtime of Lambdas

### PR DESCRIPTION
**Pull request by whit537**
Now that we have Python 3 support (#705, almost landed), we have a use case of migrating a lambda from one runtime to the other. However, currently we explicitly disallow this:

https://github.com/capitalone/cloud-custodian/blob/f5e34a723eda3c162bb4002a085a144b38f5e238/c7n/mu.py#L351

I can't quite find a reason we did this (it entered in c18b66351acc384d053ab782fcd7245705d8abd3), so I propose we remove it.

### CODE REVIEW

**Comment by whit537** [comment]
This is ready for review, @kapilt. I redid mailer (#1417) and sentry (#1420) to depend on it.

**Comment by kapilt** [comment]
the reason for the that config delete, is that other runtimes like java, nodejs, etc are explicitly not supported aka previously only one runtime supported. For supporting 2.7 and 3.6 as configurable we'll need to add an additional validate logic to policy lambda mode.

**Comment by whit537** [comment]
Rebased, was c1d7a8ee1c6b05d99e0e0b5f73798073139fbe18.

**Comment by whit537** [comment]
71c8b6bf7bf1c6033c13162e3fb8e8bd2f2e2b33 exposes runtime to policy config. Will look at constraining values for policy.data['mode']['runtime'] in next commit.

File: c7n/mu.py [hunk 1]
```diff
@@ -644,6 +643,10 @@ def role(self):
         return self.policy.data['mode'].get('role', '')
 
     @property
+    def runtime(self):
+        return self.policy.data['mode'].get('runtime', DEFAULT_RUNTIME)
```
Comment (kapilt) [inline c7n/mu.py]: that seems odd.. default runtime returns a dictionary? and explicit config returns a string?

File: c7n/mu.py [hunk 2]
```diff
@@ -43,6 +43,7 @@
 
 
 log = logging.getLogger('custodian.lambda')
+DEFAULT_RUNTIME = {2: 'python2.7', 3: 'python3.6'}[sys.version_info[0]]
```
Comment (whit537) [inline c7n/mu.py]: This construction allows for running `custodian` under, e.g., 3.5 locally and still pushing a 3.6 lambda.

Comment (kapilt) [reply c7n/mu.py]: yup, sounds good, sorry missed that index into the dict first time through, deleted original comment.

File: tests/zpill.py [hunk 3]
```diff
@@ -240,8 +241,10 @@ def assertJmes(self, expr, instance, expected):
     def cleanUp(self):
         pass
 
-    def record_flight_data(self, test_case, zdata=False):
+    def record_flight_data(self, test_case, zdata=False, vary_on_runtime=False):
```
Comment (kapilt) [inline tests/zpill.py]: this feels a little odd tbh, both on the helper addition, and on the caller, i'd rather the tests be deterministic, ie. if we want to test both runtimes, then let's test both, afaics the caller doesn't even care per se. updating the tests to just match default_runtime in their asserts would avoid the helper.

Comment (whit537) [reply tests/zpill.py]: I've replaced 71c8b6bf7bf1c6033c13162e3fb8e8bd2f2e2b33 with 9855d4dd8b6c5e3a883e5ba4f554ef54a52892ea, which uses `python2.7` as the default runtime for policy lambdas regardless of the local runtime. Better?

**Comment by whit537** [comment]
Rebased, was 9855d4dd8b6c5e3a883e5ba4f554ef54a52892ea.

**Comment by whit537** [comment]
> Will look at constraining values for `policy.data['mode']['runtime']` in next commit.

Done in 3068ac2a0c010c2cca739d0618c4de03196feb6e. Back over to you @kapilt.

**Pull request merged**